tests/unit/apps/test_core.py

```diff
@@ -329,6 +329,17 @@ def test_application_unknown_source(status, model, source_value, apps_machines):
         app.apt_source_codename
 
 
+@patch("cou.utils.openstack.OpenStackCodenameLookup.find_compatible_versions", return_value=[])
+def test_application_get_latest_os_version_failed(config, status, model, apps_machines):
+    app_name = "my_keystone"
+    exp_error = f"'{app_name}' with workload version 17.0.1 has no compatible OpenStack release."
+    app_status = status["keystone_focal_ussuri"]
+    app_config = config["openstack_ussuri"]
+
+    with pytest.raises(ApplicationError, match=exp_error):
+        Keystone(app_name, app_status, app_config, model, "keystone", apps_machines["keystone"])
```

Comment:
nit: I think this test would be m ore clear if you change the workload version of one unit to something that we don't consider like `15.0.1`. I took some time to understand why it was failing (mocked find_compatible_versions returning empty) because everything look right.